cdm/src/main/java/thredds/crawlabledataset/CrawlableDatasetAmazonS3.java

```diff
@@ -34,38 +40,41 @@ public CrawlableDatasetAmazonS3(String path, Object configObject)
 
     private CrawlableDatasetAmazonS3(CrawlableDatasetAmazonS3 parent, ThreddsS3Object s3Object)
     {
-        this(S3Helper.concat(parent.getPath(), s3Object.key), null);
+        this(helper.concat(parent.getPath(), s3Object.key), null);
         this.s3Object = s3Object;
     }
 
-    public static Cache getS3Cache(String cacheName) {
-        return getS3Cache(cacheName, null);
+    public static Ehcache getS3Cache(String cacheName, CacheEntryFactory factory) {
+        return getS3Cache(cacheName, factory, null);
     }
 
-    public static Cache getS3Cache(String cacheName, CacheEventListener eventListener)
+    public static synchronized Ehcache getS3Cache(String cacheName, CacheEntryFactory factory, CacheEventListener eventListener)
     {
         CacheManager cacheManager = CacheManager.create();
 
         if (!cacheManager.cacheExists(cacheName)) {
             Cache newCache = new Cache(cacheName, EHCACHE_MAX_OBJECTS, false, false, EHCACHE_TTL, EHCACHE_TTI);
+            cacheManager.addCache(newCache);
+            SelfPopulatingCache selfPopulatingCache = new SelfPopulatingCache(newCache, factory);
 
-            if (null != eventListener)
-                newCache.getCacheEventNotificationService().registerListener(eventListener);
+            if (null != eventListener) {
+                selfPopulatingCache.getCacheEventNotificationService().registerListener(eventListener);
+            }
 
-            cacheManager.addCache(newCache);
+            cacheManager.replaceCacheWithDecoratedCache(newCache, selfPopulatingCache);
         }
 
-        return cacheManager.getCache(cacheName);
+        return cacheManager.getEhcache(cacheName);
     }
 
-    private Cache getS3ObjectCache()
+    private Ehcache getS3ObjectCache()
     {
-        return getS3Cache(EHCACHE_S3_OBJECT_KEY, new S3CacheEventListener());
+        return getS3Cache(EHCACHE_S3_OBJECT_KEY, new S3ObjectCacheEntryFactory(), new S3CacheEventListener());
```

Comment:
Perhaps specify the class instead and instantiate if required.   But seems a bit ugly.